core/src/main/java/hudson/model/Run.java

```diff
@@ -2302,6 +2302,9 @@ public EnvVars getEnvironment() throws IOException, InterruptedException {
         for (EnvironmentContributor ec : EnvironmentContributor.all().reverseView())
             ec.buildEnvironmentFor(this,env,listener);
 
+        for (EnvironmentContributingAction a : getActions(EnvironmentContributingAction.class))
+            a.buildEnvVars(this,env,n);
```

Comment:
I agree. I was thinking about adding a wrapper for it using the Java 8 magic.
The only concern is about performance requirements for particular extensions.